profiling/paper_timing_tables/timing_comparison.py

```diff
@@ -0,0 +1,358 @@
+    )
+
+    # time the individual PINT functions
+    importtimes = []
+    getTOAs_nopickle = []
+    getTOAs_pickle = []
+    fittimes = []
+
+    # time import statements
+    total = 0
+    for i in range(MAXIT):
+        start = timeit.default_timer()
+        subprocess.call(
+            "python3 import_statements.py",
+            shell=True,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+        )
+        end = timeit.default_timer()
+        total = total + (end - start)
+    for i in range(len(ntoas_simple)):
+        importtimes.append(total / MAXIT)
+
+    # setup
+    m = models.get_model("NGC6440E.par")
+    use_planets = False
+    if m.PLANET_SHAPIRO.value:
+        use_planets = True
+    model_ephem = "DE421"
+    if m.EPHEM is not None:
+        model_ephem = m.EPHEM.value
+
+    for tim in timfiles:
+        # no pickle time of get_TOAs
+        print("timing get_TOAs w/o pickling...")
+        total = 0
+        for i in range(MAXIT):
+            start = timeit.default_timer()
+            toa.get_TOAs(tim, planets=use_planets, ephem=model_ephem, usepickle=False)
+            end = timeit.default_timer()
+            total = total + (end - start)
+        getTOAs_nopickle.append(total / MAXIT)
+
+        t = toa.get_TOAs(
+            tim, planets=use_planets, ephem=model_ephem, usepickle=True
+        )  # to use in timing fitter
+
+        f = fitter.WLSFitter(t, m)
+
+        print("timing fitter...")
+        total = 0
+        for i in range(MAXIT):
+            start = timeit.default_timer()
+            f.fit_toas()
+            end = timeit.default_timer()
+            total = total + (end - start)
+        fittimes.append(total / MAXIT)
+
+        # pickle time of get_TOAs
+        print("timing get_TOAs w/ pickling...")
+        total = 0
+        for i in range(MAXIT):
+            start = timeit.default_timer()
+            toa.get_TOAs(tim, planets=use_planets, ephem=model_ephem, usepickle=True)
+            end = timeit.default_timer()
+            total = total + (end - start)
+        getTOAs_pickle.append(total / MAXIT)
+
+    # create table 8 in PINT paper
+    function_comparison = Table(
+        (ntoas_simple, importtimes, getTOAs_nopickle, getTOAs_pickle, fittimes),
+        names=(
+            "Number of TOAs",
+            "Import Statements (sec)",
+            "Load TOAs - No Pickle (sec)",
+            "Load TOAs - Pickle (sec)",
+            "WLS Fitting - No Pickle (sec)",
+        ),
+    )
+    ascii.write(
+        function_comparison,
+        "function_tables.pdf",
+        Writer=ascii.Latex,
+        latexdict={"tabletype": "table*"},
+        overwrite=True,
+    )
+    """
+    # explore more complex model
+    # use J1910+1256 with the following parameter additions to the par file (to ensure GLS fit with Tempo2):
+    # - TNRedAmp -14.227505410948254
+    # - TNRedGam 4.91353
+    # - TNRedC 45
+
+    # add needed params for GLS fitting
```

Comment:
Why are you using `subprocess` and `echo` and modifying these files? Better to modify them with python - for example, open them with append mode and just write the new line. What happens if this is run multiple times? Are these files created afresh every time this script is run?

Reply:
Thank you for all the helpful tips @aarchiba , sorry it took me so long to reply! I will be deleting the files at the conclusion of the run, I failed to delete them all in this current version but I am working on implementing better functionality right now.